src/gc-pages.c

```diff
@@ -56,231 +55,59 @@ static char *jl_gc_try_alloc_pages(int pg_cnt) JL_NOTSAFEPOINT
-                            assert(info.meta->data);
-                            // new pages available starting at min of lb and pagetable_i32
-                            if (memory_map.lb < info.pagetable_i32)
-                                memory_map.lb = info.pagetable_i32;
-                            if (info.pagetable1->lb < info.pagetable1_i32)
-                                info.pagetable1->lb = info.pagetable1_i32;
-                            if (info.pagetable0->lb < info.pagetable0_i32)
-                                info.pagetable0->lb = info.pagetable0_i32;
-                            goto have_free_page; // break out of all of these loops
-                        }
-                    }
-                    info.pagetable1->freemap0[info.pagetable1_i32] &= ~(uint32_t)(1u << info.pagetable1_i); // record that this was full
-                }
-            }
-            memory_map.freemap1[info.pagetable_i32] &= ~(uint32_t)(1u << info.pagetable_i); // record that this was full
-        }
-    }
+    jl_gc_pagemeta_t *meta = NULL;
+    jl_mutex_lock_nogc(&global_page_pool.lock);
 
-    // no existing pages found, allocate a new one
-    {
-        jl_gc_pagemeta_t *meta = jl_gc_alloc_new_page();
-        info = page_metadata_ext(meta->data);
-        assert(meta == info.meta);
-        // new pages are now available starting at max of lb and pagetable_i32
-        if (memory_map.lb > info.pagetable_i32)
-            memory_map.lb = info.pagetable_i32;
-        if (info.pagetable1->lb > info.pagetable1_i32)
-            info.pagetable1->lb = info.pagetable1_i32;
-        if (info.pagetable0->lb > info.pagetable0_i32)
-            info.pagetable0->lb = info.pagetable0_i32;
+    if (global_page_pool.page_metadata_back != NULL) {
+        meta = pop_page_metadata_back(&global_page_pool.page_metadata_back);
+        jl_mutex_unlock_nogc(&global_page_pool.lock);
+        gc_alloc_map_set(meta->data, 1);
     }
 
-have_free_page:
-    // in-use pages are now ending at min of ub and pagetable_i32
-    if (memory_map.ub < info.pagetable_i32)
-        memory_map.ub = info.pagetable_i32;
-    if (info.pagetable1->ub < info.pagetable1_i32)
-        info.pagetable1->ub = info.pagetable1_i32;
-    if (info.pagetable0->ub < info.pagetable0_i32)
-        info.pagetable0->ub = info.pagetable0_i32;
-
-    // mark this entry as in-use and not free
-    info.pagetable0->freemap[info.pagetable0_i32] &= ~(uint32_t)(1u << info.pagetable0_i);
-    info.pagetable0->allocmap[info.pagetable0_i32] |= (uint32_t)(1u << info.pagetable0_i);
-    info.pagetable1->allocmap0[info.pagetable1_i32] |= (uint32_t)(1u << info.pagetable1_i);
-    memory_map.allocmap1[info.pagetable_i32] |= (uint32_t)(1u << info.pagetable_i);
-
-#ifdef _OS_WINDOWS_
-    VirtualAlloc(info.meta->data, GC_PAGE_SZ, MEM_COMMIT, PAGE_READWRITE);
-#endif
+    if (meta == NULL) {
+        jl_mutex_unlock_nogc(&global_page_pool.lock);
+        char *data = jl_gc_try_alloc_pages();
```

Comment:
This should possibly happen behind the lock, so we don't accidentally decide to allocate 64MB*N threads all at once. On newer systems with 64+ cores, that could be upwards of 4GB we try to reserve here all at once.